code/parse/sexp_container.cpp

```diff
@@ -92,14 +124,28 @@ void init_sexp_containers()
 	Containers_by_name_map.clear();
 }
 
-void update_sexp_containers(SCP_vector<sexp_container> &containers)
+void update_sexp_containers(SCP_vector<sexp_container>& containers,
+	const SCP_unordered_map<SCP_string, SCP_string, SCP_string_lcase_hash, SCP_string_lcase_equal_to>
+		&renamed_containers)
 {
 	Sexp_containers = std::move(containers);
 
 	Containers_by_name_map.clear();
 	for (auto &container : Sexp_containers) {
 		Containers_by_name_map.emplace(container.container_name, &container);
 	}
+
+	if (!renamed_containers.empty()) {
+		for (int i = 0; i < Num_sexp_nodes; i++) {
+			auto &node = Sexp_nodes[i];
+			if (node.type == SEXP_ATOM && node.subtype == SEXP_ATOM_CONTAINER) {
+				const auto new_name_it = renamed_containers.find(node.text);
+				if (new_name_it != renamed_containers.cend()) {
+					strcpy_s(node.text, new_name_it->second.c_str());
```

Comment:
Copying something directly into `Sexp_nodes[].text` doesn't smell right.  I think I need to know more about how the design works.

Reply:
The reason is that the FREDder can rename containers in FRED's "Add/Modify Container" dialog. Since the text for container nodes is the container name, if the FREDder renames a container, then the SEXP nodes that refer to that container need to have their text changed.

I wonder how FRED/FSO handle the analogous issue of renaming variables. A quick test with FRED shows that when I rename a variable, references to it in the tree have their names updated. Are the variables' names not stored in the SEXP nodes' `text` field?

Reply:
I'll have to check how renaming variables is implemented.

Reply:
Ok, found it. This code is doing the container equivalent of `sexp_tree::modify_sexp_tree_variable()`.

The differences are
- You can rename multiple containers while using the Add/Modify Container dialog, but AFAICT the Modify Variable dialog allows you to rename only one variable at a time.
- With variables, any of the type/name/default value can chance, so the node needs to be fully updated. However, containers don't have default values, and their type info is stored with the container objects in `Sexp_containers`, so the only value in the tree node that needs to be updated is the name.

Let me know if that clarifies things.

Reply:
Oh wait, now I get your concern. Yeah, I think you're right, we should be updating the text in `tree_nodes` instead and then allowing FRED to update `Sexp_nodes` when it saves its view of the SEXP tree.

If that's the case, then if the FREDder makes changes in the Add/Modify Container dialog and clicks OK om that dialog *but* then presses Cancel in the Events Editor, should none of the container changes be applied to `Sexp_containers`? Does modifying variables work that way?

If so, then the current Add/Modify Container dialog behavior is incorrect, because that dialog updates `Sexp_containers` after you click OK in it, regardless of what you click in the Events Editor. I'll create a separate PR to fix that.

Reply:
Well... wait a sec. What should happen if the FREDder renames containers but then presses Cancel in the Events Editor? Will affected "Replace Container" nodes in `Sexp_nodes` get their text updated to the new container names?

I'll have to check how this situation is handled for variables.

Reply:
Interesting. When I rename a variable in the Modify Variable dialog and click OK, then click Cancel in the Events Editor, the result is an assertion failure when trying to save the mission, specifically at 
```
Assert(sexp_variable_index != -1);
```
in `CTEXT()` in the middle of `check_sexp_syntax()`.

So it looks like this (admittedly improbable) scenario isn't handled correctly for variables. I'd say that qualifies as a bug.

Reply:
> Interesting. When I rename a variable in the Modify Variable dialog and click OK, then click Cancel in the Events Editor, the result is an assertion failure when trying to save the mission, assuming that `Sexp_nodes` had a "Replace Variable" node with the affected variable prior to opening the Events Editor.
> 
> ```
> Assert(sexp_variable_index != -1);
> ```
> 
> in `CTEXT()` in the middle of `check_sexp_syntax()`.
> 
> So it looks like this (admittedly improbable) scenario isn't handled correctly for variables. I'd say that qualifies as a bug.

I vaguely remember stumbling across that bug with the original container code (which did things the same way) and then fixing it. I didn't think to check if the variable code I'd based it on also had the same bug.

Reply:
I filed Issue #3849 for the above FRED crash.

To prevent this same crash for containers, I think we should keep these `Sexp_nodes` updates.

When I add container support to FRED's SEXP tree in the next PR, then I'll also update node text in `tree_nodes` when containers are renamed.

In both cases, the text is the only value that needs to be updated when containers are modified, since all info on types/contents is stored in `Sexp_containers` itself.